### PR TITLE
feat(object): add unzipObjWith

### LINKED ISSUE #801: unzipObjWith
To form a partial isomorphism with `R.zipObj` the transformation needs to look like this:

```js
const patient = {
  'abbrev@1':{
    version:'1.0.9',
    resolved:'https://registry.yarnpkg.com/abbrev/-/abbrev-1.0.9.tgz#91b4792588a7738c25f35dd6f63752a2f8776135'
  },
  'shell-quote@git+https://github.com/srghma/node-shell-quote.git#without_unlicenced_jsonify':{
    version:'1.6.0',
    resolved:'git+https://github.com/srghma/node-shell-quote.git#0aa381896e0cd7409ead15fd444f225807a61e0a'
  },
  '@graphile/plugin-supporter@git+https://1234user:1234pass@git.graphile.com/git/users/1234user/postgraphile-supporter.git':{
    version:'1.6.0',
    resolved:'git+https://1234user:1234pass@git.graphile.com/git/users/1234user/postgraphile-supporter.git#1234commit'
  },
}

const result = unzipObjWith((value, key) => [key, { ...value, name: key }]), patient)

const expected = [
  [
    'abbrev@1', 
    'shell-quote@git+https://github.com/srghma/node-shell-quote.git#without_unlicenced_jsonify',
    '@graphile/plugin-supporter@git+https://1234user:1234pass@git.graphile.com/git/users/1234user/postgraphile-supporter.git',
 ],
 [
  {
    name: 'abbrev@1',
    version:'1.0.9',
    resolved:'https://registry.yarnpkg.com/abbrev/-/abbrev-1.0.9.tgz#91b4792588a7738c25f35dd6f63752a2f8776135'
  },
  {
    name: 'shell-quote@git+https://github.com/srghma/node-shell-quote.git#without_unlicenced_jsonify',
    version:'1.6.0',
    resolved:'git+https://github.com/srghma/node-shell-quote.git#0aa381896e0cd7409ead15fd444f225807a61e0a'
  },
  {
    name:   '@graphile/plugin-supporter@git+https://1234user:1234pass@git.graphile.com/git/users/1234user/postgraphile-supporter.git',
    version:'1.6.0',
    resolved:'git+https://1234user:1234pass@git.graphile.com/git/users/1234user/postgraphile-supporter.git#1234commit'
  },
 ]
]

expect(result).toEqual(result)
```

So you'll need to do `expected[1]` to get what you need. I like that this is generic concept which plays with the rest of the `zipping` algos.

### PR DESCRIPTION
**Pull request by thomasmichaelwallace**
Closes #801

### CODE REVIEW

File: src/unzipObjWith.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  toPairs,
+  flip,
+  transpose,
+} from 'ramda';
+
+/**
+ * Unzips an object into an two arrays of transformed keys and transformed values.
+ *
+ * When executed it converts the object into value/key pairs which are passed to the transformation
+ * functor. The returned, transformed, `[key, value]` pairs are then grouped into two arrays of
```
Comment (char0n) [inline src/unzipObjWith.js]: The description is way to complicated even for me to understand (and I know what the implementation should do). What about:
```
Creates a new list out of the supplied object by applying the function to each key/value pairing. 
```

File: src/unzipObjWith.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * @param {Function} fn The function to transform each value-key pair
+ * @param {Object} obj Object to unzip
+ * @return {Array} An array with two elements: an array of keys and an array of values
+ * @see {@link https://ramdajs.com/docs/#mapObjIndexed}
+ * @example
+ *
+ * RA.unzipObjWith((v, k) => [`new${k.toUpperCase()}`, 2 * v], { a: 1, b: 2, c: 3 });
+ * //=> [['newA', 'newB', 'newC'], [2, 4, 6]]
+ */
+const unzipObjWith = curryN(2, (fn, obj) => {
```
Comment (char0n) [inline src/unzipObjWith.js]: Can we simplify to ?
```js
const unzipObjWith = curryN(2, (fn, obj) => pipe(
  toPairs, 
  map(pipe(flip, apply, fn)),
  transpose
)(ob)))

File: test/unzipObjWith.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import { assert } from 'chai';
+import * as R from 'ramda';
+
+import * as RA from '../src';
+
+describe('unzipObjWith', function() {
+  it('unzips an object into key/value arrays applying a value/key transform', function() {
+    const obj = {
```
Comment (char0n) [inline test/unzipObjWith.js]: Pls simplify the data inside tests. The tests are so complicated here it's not apparent what is happening.

File: test/unzipObjWith.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+          resolved:
+            'git+https://1234user:1234pass@git.graphile.com/git/users/1234user/postgraphile-supporter.git#1234commit',
+        },
+      ],
+    ];
+
+    assert.deepEqual(actual, expected);
+  });
+
+  it('supports gap placeholders and currying', function() {
```
Comment (char0n) [inline test/unzipObjWith.js]: We wanna tests the full currying here. For non-unary function we have a following convention to test for currying: https://github.com/char0n/ramda-adjunct/blob/84c7066f002e0f7e7a136a8cbf80c48c9d7aea43/test/replaceAll.js#L44

Basically calling the function in every combination

File: src/unzipObjWith.js (hunk 1 above)
Comment (thomasmichaelwallace) [reply src/unzipObjWith.js]: Thanks- I really struggled to think of a good way of summing it up!

File: types/index.d.ts [hunk 5]
```diff
@@ -820,6 +820,11 @@ declare namespace RamdaAdjunct {
         flattenProp(prop: string | number, obj: object): object;
         flattenProp(prop: string | number): (obj: object) => object;
 
+        /**
+         * Creates a new list out of the supplied object by applying the function to each key/value pairing.
+         */
+        unzipObjWith<T, U, V>(fn: (value: T, key: string) => [U, V], obj: {}): [U[], V[]];
```
Comment (char0n) [inline types/index.d.ts]: Can we align signatures to look like this ?
```ts
unzipObjWith<T, U>(fn: (v: T, k: string) => [U, T], obj: { [k: string]: T }): [U[], T[]];
unzipObjWith<T, U>(fn: (v: T, k: string) => [U, T]): (obj: { [k: string]: T }) => [U[], T[]];
```
Notice the curried form of the signature.

File: types/test/unzipObjWith.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+import * as RA from 'ramda-adjunct';
+
+RA.unzipObjWith((v, t) => [t, 1], { a: 'b' }); // $ExpectType [string[], number[]]
```
Comment (char0n) [inline types/test/unzipObjWith.ts]: Let's add more tests + we're missing tests for currying

Comment (thomasmichaelwallace) [reply types/test/unzipObjWith.ts]: I don't have a huge amount of experience with typescript and type tests- I've added some more based on what I can see / understand of the other tests in the folder. But I'll need some suggestions of things to test if these don't meet what you're expecting.

File: src/unzipObjWith.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * @func unzipObjWith
+ * @memberOf RA
+ * @category Object
+ * @since {@link https://char0n.github.io/ramda-adjunct/2.22.0|v2.22.0}
+ * @sig  (v, k) => [k, v] -> { k: v } -> [[k], [v]]
+ * @param {Function} fn The function to transform each value-key pair
+ * @param {Object} obj Object to unzip
+ * @return {Array} A pair of tw lists: a list of keys and a list of values
+ * @see {@link https://ramdajs.com/docs/#mapObjIndexed}
```
Comment (char0n) [inline src/unzipObjWith.js]:
```js
@see {@link https://ramdajs.com/docs/#mapObjIndexed}, {@link RA.zipObjWith|zipObjWith}
```

**Pull request merged**